File: src/main/java/io/github/tanguygab/tabadditions/shared/Platform.java
package io.github.tanguygab.tabadditions.shared;

import dev.simplix.protocolize.api.Protocolize;
import dev.simplix.protocolize.api.inventory.PlayerInventory;
import dev.simplix.protocolize.api.item.BaseItemStack;
import dev.simplix.protocolize.api.player.ProtocolizePlayer;
import io.github.tanguygab.tabadditions.shared.features.chat.ChatItem;
import me.neznamy.tab.api.TabPlayer;
import me.neznamy.tab.api.placeholder.PlaceholderManager;
import net.kyori.adventure.audience.Audience;

import java.util.List;

public abstract class Platform {

	public abstract boolean isProxy();
	public abstract void runTask(Runnable run);
	public abstract void reload();
	public abstract void registerPlaceholders(PlaceholderManager pm);
	public abstract void registerCommand(String cmd, String... aliases);
	public abstract boolean isPluginEnabled(String plugin);
	public abstract void disable();

	public abstract void sendTitle(TabPlayer p, String title, String subtitle, int fadeIn, int stay, int fadeout);
	public abstract void sendActionbar(TabPlayer p, String text);

	public abstract Audience audience(TabPlayer player);
	public abstract void sendToDiscord(TabPlayer player, String msg, String channel, List<String> plugins);
	public abstract boolean supportsChatSuggestions();
	public abstract void updateChatComplete(TabPlayer p, List<String> emojis, boolean add);

	public ChatItem getItem(TabPlayer p, boolean offhand) {
		ProtocolizePlayer player = Protocolize.playerProvider().player(p.getUniqueId());
		PlayerInventory inv = player.proxyInventory();
		BaseItemStack item = inv.item(offhand ? inv.heldItem() : 45);
		String displayName = getItemName(item.displayName(),item.itemType().toString());
		return new ChatItem(item.itemType().toString(),displayName,item.amount(),item.nbtData().toString());
	}

	protected String getItemName(String displayName, String type) {
		if (displayName != null) return displayName;
		type = type.replace("_", " ").toLowerCase();
		StringBuilder type2 = new StringBuilder();
		List<String> typelist = List.of(type.split(" "));
		for (String str : typelist) {
			type2.append(str.substring(0, 1).toUpperCase()).append(str.substring(1));
			if (typelist.indexOf(str) != typelist.size() - 1) type2.append(" ");
		}
		return type2.toString();
	}

}

File: src/main/kotlin/io/github/tanguygab/tabadditions/shared/Platform.kt
package io.github.tanguygab.tabadditions.shared

import dev.simplix.protocolize.api.Protocolize
import io.github.tanguygab.tabadditions.shared.features.chat.ChatItem
import me.neznamy.tab.api.TabPlayer
import me.neznamy.tab.api.placeholder.PlaceholderManager
import net.kyori.adventure.audience.Audience

abstract class Platform {
    abstract val isProxy: Boolean
    abstract fun runTask(run: () -> Unit)
    abstract fun reload()
    abstract fun registerPlaceholders(pm: PlaceholderManager)
    abstract fun isPluginEnabled(plugin: String): Boolean
    abstract fun disable()

    abstract fun sendTitle(player: TabPlayer, title: String, subtitle: String, fadeIn: Int, stay: Int, fadeout: Int)
    abstract fun sendActionbar(player: TabPlayer, text: String)

    abstract fun audience(player: TabPlayer?): Audience
    abstract fun sendToDiscord(player: TabPlayer, msg: String, channel: String, plugins: List<String>)
    abstract fun supportsChatSuggestions(): Boolean
    abstract fun updateChatComplete(player: TabPlayer, emojis: List<String>, add: Boolean)

    open fun getItem(player: TabPlayer, offhand: Boolean): ChatItem {
        val player = Protocolize.playerProvider().player(player.uniqueId)
        val inv = player.proxyInventory()
        val item = inv.item(if (offhand) inv.heldItem().toInt() else 45)
        val displayName = getItemName(item.displayName().asLegacyText(), item.itemType().toString())
        return ChatItem(item.itemType().toString(), displayName, item.amount().toInt(), item.nbtData().toString())
    }

    protected fun getItemName(displayName: String?, type: String): String {
        if (displayName != null) return displayName
        val type = type.replace("_", " ").lowercase()
        val type2 = StringBuilder()
        val typeList = type.split(" ")
        for (str in typeList) {
            type2.append(str.take(1).uppercase()).append(str.substring(1))
            if (typeList.indexOf(str) != typeList.size - 1) type2.append(" ")
        }
        return type2.toString()
    }
}